### ISSUE TITLE
restic forget: Keep also oldest snapshot when there are not enough backups to keep 

### ISSUE THREAD

**Issue by JsBergbau** [restic/restic#3624]
Output of `restic version`
--------------------------
restic 0.12.1 compiled with go1.16.6 on linux/arm

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------
When forgetting snapshots and there aren't enough snapshots to keep, keep also the oldest snapshot. Especially with yearly backups it can otherwise take a long time until forget policy is working the desired way.

What are you trying to do? What problem would this solve?
---------------------------------------------------------
I'm trying to keep two yearly backups, so I can go back in time at least for one year. 

A full description together with the suggestion to open a feature request, can be found here https://forum.restic.net/t/keeping-yearly-snapshots-policy-when-backup-began-during-the-year/4670

This feature request also helps losing data. As in the forum described I would have lost my oldest backup because restic forget works in this example different than expected. In my opinion it is not intuitive the way it currently works and you must take care not to delete too much snapshots. This feature request prevents this.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
Restic helps me everyday because it makes me feel safe, that my data is safe when I accidentially delete something or HDD is damaged. I've tried a lot backup solution including recovery process. None of them was really convincing. Restic is the first backup solution I'm really happy with and there is also so much potential to make restic even better.

**Comment by whereisaaron** [comment]
Just checking you're aware you can stack 'keep' options. So e.g. if you're doing daily back-ups, you could start with e.g.

```
--keep-monthly 12 --keep-yearly 2
```

Which would keep one per month for a year, then those one of those can graduate into keeping two yearly backups.

**Comment by JsBergbau** [comment]
Yeah I'm aware of that. That solution was also suggested at the forum link from the first post.
Still this problem also applies when beginning within a month, thats why this feature request is made. Details see the linked forum thread. Beeing unaware of this behaviour could lead to deleting too much snapshots which should be avoided.

**Comment by lima1** [comment]
I just started using restic and I think I ran into the same issue - or I'm doing it wrong. I put

```
restic backup --files-from=/root/restic.files --tag automated
restic forget --keep-within-daily 7d --keep-within-weekly 1m --keep-within-monthly 1y --keep-within-yearly 75y --prune --cleanup-cache --dry-run

```
Into my cron.daily assuming I end up with daily backups for the last week, then 4 weekly, then 12 monthly, and then a single one for each past year, but it keeps deleting the 8th backup, so it's only the last 7 I end up with.

**Comment by gum3ng** [comment]
Hi I would like to help with this. As far as I can understand, I came up with this solution:
1. As soon as the policy is applied on the snapshots, we can run a check on the list of snapshots to be removed to ensure that if there are no enough backups, the oldest snapshots are kept
For eg : `restic forget --keep-daily 7 --keep-monthly 4 --keep-yearly 3 --keep-weekly 4 --prune`   
In the example, we can do a check to ensure that atleast 2 snapshots are kept (1 latest and 2 oldest).

I would appreciate any feedback on this.

**Comment by MichaelEischer** [comment]
The description before the example sounds right. But I'm not sure I understand the example?

**Comment by gum3ng** [comment]
The option '--keep-yearly 3' has the highest precedence. So restic should ensure that atleast 3 snapshots are kept. Is this logic correct or am I missing something?

**Comment by MichaelEischer** [comment]
What would be the expected result for `--keep-yearly 2 --keep-monthly 50` (I have to admit it's a bit nonsensical)? I think we should just keep the oldest snapshot if any of the `--keep-*` options does not collect enough different snapshots.

**Comment by JsBergbau** [comment]
>  I think we should just keep the oldest snapshot if any of the --keep-* options does not collect enough different snapshots.

Thats a really great a idea.

**Comment by MortimerCat** [comment]
I came here because I realised the folder I just added (Jan 2023) will first save as a yearly in 2024.  I keep six months but by the end of this year, I will have lost any record of today. 

This would work for me...
> > I think we should keep the oldest snapshot if any of the --keep-* options do not collect enough different snapshots.

I have come up with another, possibly more straightforward option -  How about an `--oldest` option?   This would just simply keep the first backup per server/path.
